controller/src/test/java/io/pravega/controller/eventProcessor/impl/EventProcessorTest.java

```diff
@@ -369,6 +369,58 @@ public void testEventProcessorCell() throws CheckpointStoreException, Reinitiali
         cell.awaitTerminated();
     }
 
+    @Test(timeout = 10000)
+    public void testEventProcessorCellShutdown() throws InterruptedException {
+        CheckpointStore checkpointStore = CheckpointStoreFactory.createInMemoryStore();
+
+        CheckpointConfig.CheckpointPeriod period =
+                CheckpointConfig.CheckpointPeriod.builder()
+                        .numEvents(1)
+                        .numSeconds(1)
+                        .build();
+
+        CheckpointConfig checkpointConfig =
+                CheckpointConfig.builder()
+                        .type(CheckpointConfig.Type.Periodic)
+                        .checkpointPeriod(period)
+                        .build();
+
+        EventProcessorGroupConfig config =
+                EventProcessorGroupConfigImpl.builder()
+                        .eventProcessorCount(1)
+                        .readerGroupName(READER_GROUP)
+                        .streamName(STREAM_NAME)
+                        .checkpointConfig(checkpointConfig)
+                        .build();
+
+        EventProcessorSystem system = Mockito.mock(EventProcessorSystem.class);
+        Mockito.when(system.getProcess()).thenReturn(PROCESS);
+
+        EventStreamReader<TestEvent> reader = Mockito.mock(EventStreamReader.class);
+
+        Mockito.when(reader.readNextEvent(anyLong())).thenAnswer((Answer) invocation -> {
+            //Below sleep is to simulate the scenario where infinite retries happen to connect to segmentstore when SS is down
+            Thread.sleep(60000);
+            return null;
+        });
+
+        EventProcessorConfig<TestEvent> eventProcessorConfig = EventProcessorConfig.<TestEvent>builder()
+                .supplier(() -> new TestEventProcessor(false))
+                .serializer(new EventSerializer<>())
+                .decider((Throwable e) -> ExceptionHandler.Directive.Stop)
+                .config(config)
+                .build();
+        EventProcessorCell<TestEvent> cell = new EventProcessorCell<>(eventProcessorConfig, reader,
+                new EventStreamWriterMock<>(), system.getProcess(), READER_ID, 0, checkpointStore);
+
+        cell.startAsync();
+        cell.awaitStartupComplete();
+
+        Thread.sleep(1000); //To ensure stopAsync call is made after eventProcessor run has been called
```

Comment:
Can this be made to an `assertEventuallyEquals`? That would allow it to only take as long as needed.

Reply:
Here I gave sleep of 1sec just to make sure stopAsync call is made after the eventProcessors cell's run method is called. Anyway, I have moved the stopAsync call to be invoked when reader.readNextEvent is called and removed the sleep of 1 sec.